DAG

ejholmes opened the pull request:
This is an internal refactor of `stacker.plan.Plan` to use a DAG to internally represent steps and their dependencies. The net result provides a simpler method to walk the graph, and update stacks in order of their dependencies.

Currently, this doesn't walk the graph in parallel, which is implemented in https://github.com/remind101/stacker/pull/281.

The DAG implementation is pulled from https://github.com/thieman/py-dag with the addition of a `walk` method for walking the graph in topological order.

**TODO**

* [x] Better reverse walk.

ejholmes commented:
Ran this against our staging environment without issue. Gonna take a look at adding a `walk_parallel` method to the DAG.

phobologic commented on stacker/plan.py:
```diff
@@ -97,174 +86,79 @@ def submit(self):
-
-    def __init__(self, description, sleep_time=5, wait_func=None,
-                 watch_func=None, *args, **kwargs):
+class Plan():
```
this should be `class Plan(object):`

phobologic commented on stacker/plan.py:
```diff
@@ -97,174 +86,79 @@ def submit(self):
-    def __init__(self, description, sleep_time=5, wait_func=None,
-                 watch_func=None, *args, **kwargs):
+class Plan():
+    def __init__(self, description, reverse=False, sleep_func=sleep):
```
Probably worth adding back the doc string here.

phobologic commented on stacker/plan.py:
```diff
@@ -97,174 +86,79 @@ def submit(self):
-        Returns:
-            list: A list of :class:`Step` objects that are in the given status.
+    def build(self, stacks):
+        """ Builds an internal dag from the stacks and their dependencies """
```
Add the full docstring here for the API docs.

phobologic commented on stacker/plan.py:
```diff
@@ -97,174 +86,79 @@ def submit(self):
+                try:
+                    dag.add_edge(stack.fqn, dep)
+                except DAGValidationError:
+                    raise CyclicDependencyError(stack.fqn)
```
You should add the dependency to this exception as well, just so it's easier to see what dependency caused the error.

phobologic commented on stacker/plan.py:
```diff
@@ -97,174 +86,79 @@ def submit(self):
+                    raise CyclicDependencyError(stack.fqn)
+
+        self._dag = dag
+        return None
```
You don't have to do this - 

```
>>> def blah():
...   print "WOO"
...
>>>
>>> x = blah()
WOO
>>> x
>>> type(x)
<type 'NoneType'>
```

phobologic commented on stacker/dag/__init__.py:
```diff
@@ -0,0 +1,221 @@
+        if not graph:
+            graph = self.graph
+        for node, edges in graph.items():
+
```
Weird blank line here.

phobologic commented on stacker/dag/__init__.py:
```diff
@@ -0,0 +1,221 @@
+
+    def validate(self, graph=None):
+        """ Returns (Boolean, message) of whether DAG is valid. """
+        graph = graph if graph is not None else self.graph
```
This is different than the way it's done throughout this class - might be worth changing it just for consistency.

ejholmes commented:
Ok. Think I got all the changes in based on feedback.

We should probably hold off on merging until https://github.com/remind101/stacker/pull/281 is ready, since this will be slower to start without a parallel walk.

ejholmes commented on stacker/dag/__init__.py:
```diff
@@ -0,0 +1,224 @@
+        if not graph:
+            graph = self.graph
+        nodes = self.topological_sort(graph=graph)
+        if reverse:
```
I think this is kinda a naive implementation for reversing a dag. Doing this, we can't walk the graph concurrently in reverse. Instead, we should do a depth first walk, and build a new graph with the edges reversed.

ejholmes replied:
Added a commit that adds a `tranpose` method on DAG to build a new DAG with all the edges transposed. Plan uses this now for destroy actions, so it can be parallelized.

ejholmes commented on stacker/actions/build.py:
```diff
@@ -271,26 +271,10 @@ def _handle_missing_parameters(self, params, required_params,
     def _generate_plan(self, tail=False):
-        plan_kwargs = {}
-        if tail:
-            plan_kwargs["watch_func"] = self.provider.tail_stack
```
This gets removed in this PR, which may not be correct. This seems to tail the events, but I'm not sure I've ever actually seen stacker do this.

phobologic replied:
Yeah - this is used by some people with `--tail`.  Does it cause issues if you leave it in?

phobologic commented on stacker/tests/test_plan.py:
```diff
@@ -60,20 +60,47 @@ def test_build_plan(self):
 
-        with self.assertRaises(CyclicDependencyError):
-            plan.build([vpc, bastion])
+        try:
```
I think what you want here is `self.assertRaises()`.  You can see an example of that here:

https://github.com/remind101/stacker/blob/6a141b6d9c429d89158ad7ea274cdfcfbdc6060d/stacker/tests/test_config.py#L15

ejholmes replied:
Oh sweet. I couldn't find an example online that used assertRaises to check the message. Much nicer.

phobologic commented:
Oh, also, once you submit this, could you open another PR to merge this against the `release-1.0` branch?

ejholmes commented:
Alright, this has gotten behind `master` which is moving towards 1.0, and my original version of this was based on top of 0.8.5. I've updated this for 0.8.6, if anybody wants to use it ([dag-0.8.6 branch](https://github.com/remind101/stacker/tree/dag-0.8.6), much faster than 0.8.6), but it won't be merged into 0.8 stable.

I'll follow up with porting this forward to 1.0 once we've moved internally to 1.0.
